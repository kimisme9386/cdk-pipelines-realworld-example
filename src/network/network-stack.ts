import * as ec2 from '@aws-cdk/aws-ec2';
import { Construct, Stack, StackProps, Fn } from '@aws-cdk/core';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as cdk from '@aws-cdk/core';

export interface StageConfig {
  vpc: {
      maxAzs: number;
      natGateways: number;
      ipv6enabled: boolean;
  };
  alb: {
      listener: {
      certificateArn: string;
      };
      targetHealthCheck: {
      enabled: boolean;
      interval: number;
      path: string;
      timeout: number;
      healthyThresholdCount: number;
      unhealthyThresholdCount: number;
      };
      deregistrationDelay: number;
  };
}
  

interface NetworkStackProps extends StackProps {
  stageConfig: StageConfig;
}

export class NetworkStack extends Stack {
  vpc: ec2.Vpc;
  alb: elbv2.ApplicationLoadBalancer;


  constructor(scope: Construct, id: string, props: NetworkStackProps) {
    super(scope, id, props);

    const subnetConfiguration = [
        {
          cidrMask: 24,
          name: 'ingress',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 28,
          name: 'rds',
          subnetType: ec2.SubnetType.ISOLATED,
        },
      ];
  
      if (props.stageConfig.vpc.natGateways != 0) {
        subnetConfiguration.push({
          cidrMask: 24,
          name: 'application',
          subnetType: ec2.SubnetType.PRIVATE,
        });
      }
  
      this.vpc = new ec2.Vpc(this, 'Vpc', {
        cidr: '10.0.0.0/16',
        maxAzs: props.stageConfig.vpc.maxAzs,
        natGateways: props.stageConfig.vpc.natGateways,
        subnetConfiguration: subnetConfiguration,
      });
  
      if (props.stageConfig.vpc.ipv6enabled) this.enableIpv6(this.vpc);
  
      this.alb = new elbv2.ApplicationLoadBalancer(this, 'Alb', {
        vpc: this.vpc,
        internetFacing: true,
        ipAddressType: elbv2.IpAddressType.DUAL_STACK,
        vpcSubnets: this.vpc.selectSubnets({
          subnetType: ec2.SubnetType.PUBLIC,
        }),
      });
    }
  
    enableIpv6(vpc: ec2.Vpc) {
      const cfnIpv6Cidr = new ec2.CfnVPCCidrBlock(this, 'Ipv6Cidr', {
        vpcId: vpc.vpcId,
        amazonProvidedIpv6CidrBlock: true,
      });
  
      this.vpc.publicSubnets.forEach((subnet, idx) => {
        const vpcCidrBlock = cdk.Fn.select(0, this.vpc.vpcIpv6CidrBlocks);
        const ipv6Cidrs = cdk.Fn.cidr(
          vpcCidrBlock,
          this.vpc.publicSubnets.length,
          '64'
        );
        const cfnSubnet = subnet.node.defaultChild as ec2.CfnSubnet;
        cfnSubnet.ipv6CidrBlock = cdk.Fn.select(idx, ipv6Cidrs);
        cfnSubnet.addDependsOn(cfnIpv6Cidr);
      });
  
      cfnIpv6Cidr.addDependsOn(this.vpc.node.defaultChild as ec2.CfnVPC);
    }
}