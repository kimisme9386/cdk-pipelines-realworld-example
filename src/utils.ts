import * as yaml from 'js-yaml';
export function readConfig(configPath: string): any {
  return yaml.safeLoad(configPath);
}