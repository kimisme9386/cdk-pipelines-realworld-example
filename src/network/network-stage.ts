import * as fs from 'fs';
import { Construct, Stage, StageProps } from '@aws-cdk/core';
import { readConfig } from '../utils';
import { NetworkStack, StageConfig } from './network-stack';

enum StageEnv {
  DEV = 'dev',
  PROD = 'prod',
}

interface NetworkStageProps extends StageProps {
  stageEnv: string;
}

export class NetworkStage extends Stage {
  private _networkStack:NetworkStack;

  constructor(scope: Construct, id: string, props: NetworkStageProps) {
    super(scope, id, props);

    const stageEnv = props?.stageEnv;

    if (stageEnv === undefined || !Object.values(StageEnv).includes(stageEnv as StageEnv)) {
      throw new Error('stageEnv is required');
    }

    const stageConfig: StageConfig = readConfig(
      fs.readFileSync(`${__dirname}/configs/${stageEnv}.yml`, 'utf8'),
    ) as StageConfig;

    this._networkStack = new NetworkStack(this, 'Network', {
      stageConfig,
      env: props?.env,
    });
  }

  public get networkStack() {
    return this._networkStack;
  }
}