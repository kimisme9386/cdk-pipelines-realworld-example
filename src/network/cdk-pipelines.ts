import * as codebuild from '@aws-cdk/aws-codebuild';
import { Construct, Stack, StackProps } from '@aws-cdk/core';
import * as pipelines from '@aws-cdk/pipelines';
import { NetworkStage } from './network-stage';

interface CdkPipelinesProps extends StackProps {
  cdkVersion: string;
  devAccount: string;
  stagingAccount: string;
  prodAccount: string;
}

/**
 * The stack that defines the application pipeline
 */
export class CdkPipelines extends Stack {
  constructor(scope: Construct, id: string, props: CdkPipelinesProps) {
    super(scope, id, props);

    const codePipelineSource = pipelines.CodePipelineSource.connection(
      'wondercore-devops/wondercise-service-infra',
      'main',
      {
        connectionArn:
          'arn:aws:codestar-connections:ap-northeast-1:794029059620:connection/2f474bca-8f43-41e6-b3ad-b36f3e4c5d4d',
      },
    );

    const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
      // The pipeline name
      pipelineName: 'NetworkPipelines',

      // How it will be built and synthesized
      synth: new pipelines.ShellStep('Synth', {
        // Where the source can be found
        input: codePipelineSource,

        // Install dependencies, build and run cdk synth
        commands: [
          `npm install -g aws-cdk@${props.cdkVersion}`,
          'yarn install --frozen-lockfile',
          'npx projen',
          'cdk synth',
        ],
      }),
      crossAccountKeys: true,
      selfMutationCodeBuildDefaults: {
        partialBuildSpec: codebuild.BuildSpec.fromObject({
          phases: {
            pre_build: {
              commands: [
                `npm install -g aws-cdk@${props.cdkVersion}`,
              ],
            },
          },
        }),
      },
    });

    const envWave = pipeline.addWave('Network');

    envWave.addStage(
      new NetworkStage(this, 'Dev', {
        env: { account: props.devAccount, region: 'ap-northeast-1' },
        stageEnv: 'dev',
      }),
      {
        pre: [new pipelines.ShellStep('Validate dev CloudFormation Synth', {
          commands: [
            'yarn install --frozen-lockfile',
            './node_modules/.bin/jest --passWithNoTests test/network-dev.test.ts',
          ],
        })],
      },
    );

    envWave.addStage(
      new NetworkStage(this, 'Production', {
        env: { account: props.stagingAccount, region: 'ap-northeast-1' },
        stageEnv: 'prod',
      }),
      {
        pre: [new pipelines.ShellStep('Validate staging CloudFormation Synth', {
          commands: [
            'yarn install --frozen-lockfile',
            './node_modules/.bin/jest --passWithNoTests test/network-staging.test.ts',
          ],
        })],
      },
    );
  }
}
