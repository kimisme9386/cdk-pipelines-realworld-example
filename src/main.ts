import * as fs from 'fs';
import { App } from '@aws-cdk/core';
import { CdkPipelines as EcsCdkPipelines } from './ecs/cdk-pipelines';
import { CdkPipelines as NetworkCdkPipelines } from './network/cdk-pipelines';
import { readConfig } from './utils';


export interface GlobalConfig {
  cdkVersion: string;
  account: {
    dev: string;
    prod: string;
  };
}

const app = new App();

const globalConfig: GlobalConfig = readConfig(
  fs.readFileSync(`${__dirname}/../config.yml`, 'utf8'),
) as GlobalConfig;

new NetworkCdkPipelines(app, 'TestNetwork', {
  env: {
    account: '794029059620',
    region: 'ap-northeast-1',
  },
  cdkVersion: globalConfig.cdkVersion,
  devAccount: globalConfig.account.dev,
  prodAccount: globalConfig.account.prod,
});

new EcsCdkPipelines(app, 'TestEcs', {
  env: {
    account: '794029059620',
    region: 'ap-northeast-1',
  },
  cdkVersion: globalConfig.cdkVersion,
  devAccount: globalConfig.account.dev,
  prodAccount: globalConfig.account.prod,
});

app.synth();