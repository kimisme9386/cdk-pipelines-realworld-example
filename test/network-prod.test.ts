import '@aws-cdk/assert/jest';
import * as fs from 'fs';
import { SynthUtils } from '@aws-cdk/assert';
import { App } from '@aws-cdk/core';
import { GlobalConfig } from '../src/main';
import { NetworkStage } from '../src/network/network-stage';
import { readConfig } from '../src/utils';

test('EcsApiStage Snapshot', () => {
  const app = new App();

  const globalConfig: GlobalConfig = readConfig(
    fs.readFileSync(`${__dirname}/../config.yml`, 'utf8'),
  ) as GlobalConfig;

  const networkStage = new NetworkStage(app, 'prod', {
    env: { account: globalConfig.account.dev, region: 'ap-northeast-1' },
    stageEnv: 'prod',
  });

  expect(SynthUtils.toCloudFormation(networkStage.networkStack)).toMatchSnapshot();
});