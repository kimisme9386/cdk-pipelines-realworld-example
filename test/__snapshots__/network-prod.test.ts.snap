// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcsApiStage Snapshot 1`] = `
Object {
  "Resources": Object {
    "Alb16C2F182": Object {
      "DependsOn": Array [
        "VpcingressSubnet1DefaultRoute4188A546",
        "VpcingressSubnet2DefaultRoute18203BC3",
      ],
      "Properties": Object {
        "IpAddressType": "dualstack",
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "AlbSecurityGroup580F65A6",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          Object {
            "Ref": "VpcingressSubnet1Subnet059F22C6",
          },
          Object {
            "Ref": "VpcingressSubnet2SubnetF175D754",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "AlbSecurityGroup580F65A6": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB prodNetworkAlb1DA0DB68",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "Ipv6Cidr": Object {
      "DependsOn": Array [
        "Vpc8378EB38",
      ],
      "Properties": Object {
        "AmazonProvidedIpv6CidrBlock": true,
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCCidrBlock",
    },
    "Vpc8378EB38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcIGWD7BA715C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcVPCGWBF912B6E": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "VpcingressSubnet1DefaultRoute4188A546": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcingressSubnet1RouteTable804C7A26",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcingressSubnet1RouteTable804C7A26": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/ingressSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcingressSubnet1RouteTableAssociation5DF54E70": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcingressSubnet1RouteTable804C7A26",
        },
        "SubnetId": Object {
          "Ref": "VpcingressSubnet1Subnet059F22C6",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcingressSubnet1Subnet059F22C6": Object {
      "DependsOn": Array [
        "Ipv6Cidr",
      ],
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.0.0/24",
        "Ipv6CidrBlock": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::Cidr": Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::GetAtt": Array [
                        "Vpc8378EB38",
                        "Ipv6CidrBlocks",
                      ],
                    },
                  ],
                },
                2,
                "64",
              ],
            },
          ],
        },
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/ingressSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcingressSubnet2DefaultRoute18203BC3": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcingressSubnet2RouteTable8B0E23A7",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcingressSubnet2RouteTable8B0E23A7": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/ingressSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcingressSubnet2RouteTableAssociation39E1BA3A": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcingressSubnet2RouteTable8B0E23A7",
        },
        "SubnetId": Object {
          "Ref": "VpcingressSubnet2SubnetF175D754",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcingressSubnet2SubnetF175D754": Object {
      "DependsOn": Array [
        "Ipv6Cidr",
      ],
      "Properties": Object {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.1.0/24",
        "Ipv6CidrBlock": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::Cidr": Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::GetAtt": Array [
                        "Vpc8378EB38",
                        "Ipv6CidrBlocks",
                      ],
                    },
                  ],
                },
                2,
                "64",
              ],
            },
          ],
        },
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/ingressSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcrdsSubnet1RouteTableAssociationE01668F2": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcrdsSubnet1RouteTableE62E4ED6",
        },
        "SubnetId": Object {
          "Ref": "VpcrdsSubnet1SubnetC2926CEA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcrdsSubnet1RouteTableE62E4ED6": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/rdsSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcrdsSubnet1SubnetC2926CEA": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.2.0/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "rds",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/rdsSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcrdsSubnet2RouteTable3E531D9B": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/rdsSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcrdsSubnet2RouteTableAssociation25A7BD68": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcrdsSubnet2RouteTable3E531D9B",
        },
        "SubnetId": Object {
          "Ref": "VpcrdsSubnet2Subnet70A835C8",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcrdsSubnet2Subnet70A835C8": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.2.16/28",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "rds",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "prod/Network/Vpc/rdsSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;
